### PR TITLE
En 6078 genesis file format

### PR DESCRIPTION
**Pull request by iulianpascalau**
genesis file parser + tests

### CODE REVIEW

File: genesis/data_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		createStakingGenesisEntry(),
+	}
+
+	buff, err := json.MarshalIndent(initialBalances, "", "    ")
+	assert.Nil(t, err)
+
+	fmt.Println(string(buff))
+
+	recoveredInitialBalances := make([]InitialBalance, 0)
+	err = json.Unmarshal(buff, &recoveredInitialBalances)
```
Comment (golangcibot) [inline genesis/data_test.go:28]: ineffectual assignment to `err` (from `ineffassign`)

File: genesis/data.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package genesis
+
+import "math/big"
+
+// DelegationData specify the delegation address and the balance provided
+type DelegationData struct {
+	Address string `json:"address,omitempty"`
+	Balance string `json:"balance,omitempty"`
```
Comment (sasurobert) [inline genesis/data.go:8]: I would rename this to VALUE

File: genesis/genesis.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+	if entireSupply == nil {
+		return nil, ErrNilEntireSupply
+	}
+	if big.NewInt(0).Cmp(entireSupply) >= 0 {
+		return nil, ErrInvalidEntireSupply
+	}
+
+	initialBalances := make([]*InitialBalance, 0)
+	err := core.LoadJsonFile(&initialBalances, genesisFilePath)
```
Comment (sasurobert) [inline genesis/genesis.go:36]: this will be hard to test :dagger:

Comment (iulianpascalau) [reply genesis/genesis.go:36]: Agree, added testdata files as to check (also) the happy path behavior as long as some edge cases: file not found or a file containing a bad config. Code coverage is at 100% in this file.

File: genesis/testdata/genesis_bad.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    "address": "not-an-address",
+    "supply": "1",
+    "balance": "1",
+    "stakingbalance": "0",
+    "delegation": {
+      "address": "",
+      "balance": "0"
+    }
+  }
+]
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply genesis/testdata/genesis_bad.json:12]: done

File: genesis/testdata/genesis_ok.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    "address": "0008",
+    "supply": "5",
+    "balance": "1",
+    "stakingbalance": "2",
+    "delegation": {
+      "address": "0009",
+      "balance": "2"
+    }
+  }
+]
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply genesis/testdata/genesis_ok.json:62]: done

File: genesis/data.go (hunk 2 above)
Comment (iulianpascalau) [reply genesis/data.go:8]: done

File: genesis/data.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+// InitialBalance provides information about one entry in the genesis file
+type InitialBalance struct {
+	Address        string          `json:"address"`
+	Supply         string          `json:"supply"`
+	Balance        string          `json:"balance"`
+	StakingBalance string          `json:"stakingbalance"`
+	Delegation     *DelegationData `json:"delegation"`
+
+	address        []byte
```
Comment (ccorcoveanu) [inline genesis/data.go:21]: These fields are just so we would not use separate structures for decoding the json file? Wouldn't a custom unmarshal function be cleaner?

File: genesis/genesis.go [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+			ErrEntireSupplyMismatch,
+			g.entireSupply.String(),
+			totalSupply.String(),
+		)
+	}
+
+	return nil
+}
+
+func (g *Genesis) tryParseElement(initialBalance *InitialBalance) error {
```
Comment (ccorcoveanu) [inline genesis/genesis.go:86]: Name without `try`, method can return error, so `try` is implied

File: genesis/data.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	Address string `json:"address,omitempty"`
+	Balance string `json:"balance,omitempty"`
+	address []byte
+	balance *big.Int
+}
+
+// InitialBalance provides information about one entry in the genesis file
+type InitialBalance struct {
+	Address        string          `json:"address"`
+	Supply         string          `json:"supply"`
```
Comment (ccorcoveanu) [inline genesis/data.go:16]: What does the supply for an initial balance represent?

File: genesis/data.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+// DelegationData specify the delegation address and the balance provided
+type DelegationData struct {
+	Address string `json:"address,omitempty"`
+	Balance string `json:"balance,omitempty"`
+	address []byte
+	balance *big.Int
+}
+
+// InitialBalance provides information about one entry in the genesis file
+type InitialBalance struct {
```
Comment (ccorcoveanu) [inline genesis/data.go:14]: Would this be better named `InitialAccount`? An account has balance, an address... etc

File: genesis/data.go (hunk 8 above)
Comment (iulianpascalau) [reply genesis/data.go:16]: supply = balance + staking + delegation

File: genesis/data.go (hunk 6 above)
Comment (iulianpascalau) [reply genesis/data.go:21]: Partially, address can not be automatically converted by the json marshaler as we need the PubkeyConverter. Will change the *big.Int fields

Comment (iulianpascalau) [reply genesis/data.go:21]: done for big.Int, address remains the same

File: genesis/genesis.go (hunk 7 above)
Comment (iulianpascalau) [reply genesis/genesis.go:86]: done

File: genesis/data.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	Address        string          `json:"address"`
+	Supply         string          `json:"supply"`
+	Balance        string          `json:"balance"`
+	StakingBalance string          `json:"stakingbalance"`
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply genesis/data.go:18]: done

File: genesis/data.go (hunk 9 above)
Comment (iulianpascalau) [reply genesis/data.go:14]: done
